### ISSUE TITLE
Legacy plugin not compatible with vite 2.7.x

### ISSUE THREAD

**Issue by StefSchenkelaars** [ElMassimo/vite_ruby#156]
[troubleshooting section]: https://vite-ruby.netlify.app/guide/troubleshooting.html

- [x] I have tried upgrading by running `bundle update vite_ruby`.
- [x] I have read the __[troubleshooting section]__ before opening an issue.

### Description 📖
In vite 2.7, the name of the legacy-polyfills entry has changed from `"../../vite/legacy-polyfills"` to `"../../\u0000vite/legacy-polyfills"`. This filename has changed since this polyfill is a so called virtual file ([as explained in the plugin api docs](https://github.com/vitejs/vite/blob/e19ba6a98d85586d1f54e6b0eed11dcf72e6c651/docs/guide/api-plugin.md#conventions)).

This however makes the ruby code crash since it can't resolve the filename (see stacktrace below).

This change is also discussed in https://github.com/vitejs/vite/issues/6097 but as I dove into it I found that this NULL character is not a typo but a feature. So therefore I think the gem should be change to handle this correctly.

If this could be confirmed as a bug, I could work out a solution.

### Reproduction 🐞
- Repo with vite > 2.7 and the legacy plugin
- Render a page with a `vite_legacy_javascript_tag` or `vite_legacy_typescript_tag` tag

### Logs 📜

```
ActionView::Template::Error (path name contains null byte):
vite_ruby (3.0.3) lib/vite_ruby/manifest.rb:176:in `extname'
vite_ruby (3.0.3) lib/vite_ruby/manifest.rb:176:in `with_file_extension'
vite_ruby (3.0.3) lib/vite_ruby/manifest.rb:150:in `resolve_entry_name'
vite_ruby (3.0.3) lib/vite_ruby/manifest.rb:83:in `lookup'
vite_ruby (3.0.3) lib/vite_ruby/manifest.rb:70:in `lookup!'
vite_ruby (3.0.3) lib/vite_ruby/manifest.rb:22:in `path_for'
vite_rails (3.0.2) lib/vite_rails/tag_helpers.rb:22:in `vite_asset_path'
vite_plugin_legacy (3.0.0) lib/vite_plugin_legacy/tag_helpers.rb:30:in `vite_legacy_polyfill_tag'
vite_plugin_legacy (3.0.0) lib/vite_plugin_legacy/tag_helpers.rb:15:in `vite_legacy_javascript_tag'
vite_plugin_legacy (3.0.0) lib/vite_plugin_legacy/tag_helpers.rb:20:in `vite_legacy_typescript_tag'

My app => app/views/layouts/login.html.haml:31
```

### Screenshots 📷

_Provide console or browser screenshots of the problem_.

**Comment by ElMassimo** [comment]
Released `vite_ruby@3.0.4` and `vite_plugin_legacy@3.0.1`, you can upgrade manually, or run `bin/vite upgrade`.

**Comment by khalwat** [comment]
Are we sure this isn't a regression/mistake in Vite? `\u0000` is the Unicode character for `null` which makes me think it could potentially be a mistake on their end?

**Comment by StefSchenkelaars** [comment]
Well I'm not a sure but when I started on fixing this issue in vite itself, I found that there was more code surrounding this and not just a simple typo. I created [a fix within vite](https://github.com/StefSchenkelaars/vite/commit/bbf27017f2163ca0d202d6ef71fbbec90cdfcb1b) but I wasn't really sure on how to properly write a test / spec for this. If you look at [the commit that introduced the character](https://github.com/vitejs/vite/pull/5601/files#diff-e861d33a288ea61b3aee05203d1c43920c4326d904e3d41b4326adbd30a0127fR599), it is also explicitly checked in the build step. But I agree with you that the difference between the `\u0000` character and the `0` are indeed interesting. 


So in short: I'm not sure but it works 😅

**Comment by StefSchenkelaars** [comment]
@khalwat I created a PR for my commit anyway at vite. Mainly to get the discussion going there ;) 
See: https://github.com/vitejs/vite/pull/6208

**Comment by khalwat** [comment]
hmmmm you're right, this does look done on purpose: https://github.com/vitejs/vite/pull/5601/commits/3127219bfa4129538fe5448c86b7e9aa1fe4d9ea

But I wonder if something isn't amiss with, however `\0` gets translated into `\u0000` during the build process, because the latter is what ends up in the `manifest.json`

I'd think that this is what we should end up using:

```js
const polyfillId = '\0vite/legacy-polyfills'
```

Rather than:

```js
\u0000vite/legacy-polyfills
```

**Comment by khalwat** [comment]
> In vite 2.7, the name of the legacy-polyfills entry has changed from "../../vite/legacy-polyfills" to "../../\u0000vite/legacy-polyfills". This filename has changed since this polyfill is a so called virtual file (as explained in the plugin api docs).

One thing to note from the [API docs](https://github.com/vitejs/vite/blob/e19ba6a98d85586d1f54e6b0eed11dcf72e6c651/docs/guide/api-plugin.md#conventions) is that it says:

> Internally, plugins that use virtual modules should prefix the module ID with \0 while resolving the id, a convention from the rollup ecosystem. This prevents other plugins from trying to process the id (like node resolution), and core features like sourcemaps can use this info to differentiate between virtual modules and regular files. \0 is not a permitted char in import URLs so we have to replace them during import analysis. A \0{id} virtual id ends up encoded as /@id/__x00__{id} during dev in the browser. The id will be decoded back before entering the plugins pipeline, so this is not seen by plugins hooks code.

So this id prefixing is _internal_ and should be undone before it is saved out to the `manifest.json` I'd assume. Sounds like something isn't quite right with this round-trip, at least as far as productions builds with `manifest: true` are concerned.
